### PR TITLE
Milestone 2 - Introduce `StopgapGlobalState`

### PR DESCRIPTION
**Pull request by devinmorgan**
## What
This is the PR for [Milestone 2](https://www.notion.so/joinforage/Plan-for-On-Time-Delivery-of-Android-3-0-Breaking-Changes-226d8ee6f8294d2694b1bb451791960b?pvs=4#f7302e2ba6df42fc80cd67c6e36f1fa9) of the plan to ship Android v3 to GoPuff by tomorrow EOD

## Why
We will need to temporarily rely on global state (via singleton) to ship Android v3 public contract changes ASAP without getting to [all of the required internal changes](https://www.notion.so/joinforage/Android-3-0-Design-Proposal-1ee6be95544d4dc78e7a7e7c88f46bbc?pvs=4#deb5c9f46bf74da9abc18f0b5f5ea4f4).

## Test Plan
- ✅ I've added unit tests for this change. Specifically, I've added tests for the parsing session token logic
- ❌ The reviewer should NOT manually test the changes in this PR. This PR does not change any observable behavior. Those changes will be in the next PR in this stack

## How
Wait to merge this one until the next PR in this stack is ready to be merged

### CODE REVIEW

**Comment by devinmorgan** [comment]
Current dependencies on/for this PR:
* main
  * **PR #68** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/68" target="_blank">Graphite</a> 
    * **PR #93** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/93" target="_blank">Graphite</a> 
      * **PR #97** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/97" target="_blank">Graphite</a> 
        * **PR #98** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/98" target="_blank">Graphite</a> 
          * **PR #102** <a href="https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/102" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/teamforage/forage-android-sdk/102?utm_source=stack-comment).

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+enum class EnvOption {
+    DEV,
+    STAGING,
+    SANDBOX,
+    CERT,
+    PROD
+}
+
+sealed class EnvConfig(
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:13]: should this be `internal`? Will the btProxyId, be accessible by the client?

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        btProxyID = "N31FZgKpYZpo3oQ6XiM6M6",
+        btAPIKey = "key_AZfcBuKUsV38PEeYu6ZV8x",
+        vgsVaultId = "tntlqkidhc6",
+        vgsVaultType = "SANDBOXbox",
+        baseUrl = "https://api.dev.joinforage.app/",
+        ldMobileKey = "mob-03e025cb-5b4e-4d97-8685-39a22316d601",
+        ddClientToken = "pubf13cedf24ba2ad50d4b9cb0b0100bd4a",
+    )
+
+    object Staging : EnvConfig(
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:34]: can these env objects be `internal`?

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        FLAVOR = EnvOption.STAGING,
+        btProxyID = "ScWvAUkp53xz7muae7fW5p",
+        btAPIKey = "key_6B4cvpcDCEeNDYNow9zH7c",
+        vgsVaultId = "tnteykuh975",
+        vgsVaultType = "SANDBOXbox",
+        baseUrl = "https://api.STAGINGing.joinforage.app/",
```
Comment (djoksimo) [inline forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:40]: typo from refactoring across vgsVaultType and baseUrl for all envs.

**Review by djoksimo — CHANGES_REQUESTED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 3 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:40]: lmao that's embarrassing - thank gawd for code review!

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 2 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:34]: Ah yes they very much should be!

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 1 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:13]: yessir! Good catch!

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 3 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt]: address in [eb1c88b](https://github.com/teamforage/forage-android-sdk/pull/102/commits/eb1c88b6dd6a90abca84fbb47959e992b3878673)

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 2 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:34]: address in [eb1c88b](https://github.com/teamforage/forage-android-sdk/pull/102/commits/eb1c88b6dd6a90abca84fbb47959e992b3878673)

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 1 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt]: address in [eb1c88b](https://github.com/teamforage/forage-android-sdk/pull/102/commits/eb1c88b6dd6a90abca84fbb47959e992b3878673)

**Review by djoksimo — APPROVED** [review]

**Review by dleis612 — APPROVED** [review]

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 1 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt]: somehow I only made the change for STAGINGing and not SANDBOXbox. Will update this per with the typo fix for SANDBOXbox as well

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 2 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:34]: ACTUALLY address in [ac2f223](https://github.com/teamforage/forage-android-sdk/pull/102/commits/ac2f2237d7550b346f257b88fc2061f9ab8094c8). Gonna mark as resolved!

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 3 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt]: ACTUALLY address in [ac2f223](https://github.com/teamforage/forage-android-sdk/pull/102/commits/ac2f2237d7550b346f257b88fc2061f9ab8094c8). Gonna mark as resolved!

File: forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt (hunk 2 above)
Comment (devinmorgan) [reply forage-android/src/main/java/com/joinforage/forage/android/core/EnvConfig.kt:34]: Ah didn't realize this comment was about the internal objects of the `EnvConfig`! Now that I've marked the out `EnvConfig` as `internal`, these objects will inherit this visibility as well so no need to be explicit about it here!. Gonna mark as resolved but lmk if you have any follow ups on this!

**Pull request merged**